aexpect/remote_door.py

```diff
@@ -0,0 +1,718 @@
+    uri = pyro_daemon.register(local_object)
+    if not pyrod_running:
+        loop = DaemonLoop(pyro_daemon)
+        loop.start()
+    if not nsd_running:
+        loop = DaemonLoop(ns_daemon)
+        loop.start()
+    # we should register to the name server after entering its loop
+    ns_server.register(object_name, uri)
+
+    logging.info("Local object '%s' sharing - ready", object_name)
+    loop.join()
+
+
+def share_local_objects(wait=False, host="localhost", port=0):
+    """
+    Share all local objects of the given name over the network.
+
+    :param bool wait: whether to wait for a loop exit (e.g. if not running
+                      this function's code locally)
+    :param str host: ip address of the local sharing server
+    :param int port: port of the local sharing server
+
+    This function shares all possible python code (dangerous) and not
+    just a custom object with whitelisted attributes (secure).
+    """
+    Pyro4.config.FLAME_ENABLED = True
+    Pyro4.config.SERIALIZER = "pickle"
+    Pyro4.config.SERIALIZERS_ACCEPTED = {"pickle"}
+
+    # pyro daemon
+    pyro_daemon = Pyro4.Daemon(host=host, port=port)
+    logging.debug("Pyro4 daemon started successfully")
+
+    # main retrieval of the local objects
+    from Pyro4.utils import flame
+    _uri = flame.start(pyro_daemon)
+
+    # request loop
+    loop = DaemonLoop(pyro_daemon)
+    loop.start()
+    if wait:
+        loop.join()
+
+
+def share_remote_objects(session, control_path, host="localhost", port=9090,
+                         os_type="windows", extra_params=None):
+    """
+    Create and share remote objects from a remote location over the network.
+
+    :param session: remote session to the platform
+    :type session: RemoteSession
+    :param str control_path: path to the control on the host
+    :param str host: ip address of the remote sharing server
+    :param int port: port of the remote sharing server
+    :param str os_type: OS type of the session, either "linux" or "windows"
+    :param extra_params: extra parameters to pass to the remote object sharing
+                         control file (similarly to subcontrol setting above),
+                         with keys usually prepended with "ro_" prefix
+    :type extra_params: {str, str}
+    :raises: :py:class:`RuntimeError` if the object server failed to start
+
+    In comparison to :py:func:`share_local_object`, this function fires up a
+    name server from a second spawned session (not remote util call) and uses
+    static (template) control as a remote object server which has to be
+    preprogrammed but thus also customized (e.g. sharing multiple objects).
+
+    In comparison to :py:func:`share_local_objects`, this function is not
+    an internal implementation sharing everything on the other side but only
+    what is dicated by the remote object server module and thus its creator.
+
+    .. note:: Created and works specifically for Windows and Linux.
```

Comment:
Quite impressive functionality. Thanks for taking the time to make it work well in both systems.

Reply:
And thanks a lot for reviewing! This one is definitely a tough review to go through.